processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,45 +160,44 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
-        throw new IndexSizeExceededException("Maximum number of rows reached");
-      }
-      final Integer prev = facts.putIfAbsent(key, rowOffset);
-      if (prev != null) {
-        rowOffset = prev;
-      } else {
-        Aggregator[] aggs = new Aggregator[metrics.length];
-        for (int i = 0; i < metrics.length; i++) {
-          final AggregatorFactory agg = metrics[i];
-          aggs[i] = agg.factorize(
-              makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
-          );
-        }
-        aggList.add(aggs);
-        numEntries.incrementAndGet();
+    final Aggregator[] aggs;
+    final Integer retval;
+    final Integer rowOffset = numEntries.get();
+    if (rowOffset >= maxRowCount && !facts.containsKey(key)) {
+      throw new IndexSizeExceededException("Maximum number of rows reached");
+    }
+    final Integer prev = facts.putIfAbsent(key, rowOffset);
```

Comment:
I believe the problem lies with this statement. We cannot possibly add rowOffset to the facts unless we have already created the corresponding entry in aggList. When running a query, Druid will iterate through all values of the facts table, and we don't want to introduce a read/write lock to synchronize the reading thread, instead we should always ensure that any value that exists in facts is a valid offset into aggList.

In 0.6 we always create the list of aggregators before appending it to the facts table, so the same issue can never happen.